Build docs in CI.

kousu opened the pull request:
Add GH_PAGES_TOKEN in prep to port publishing to Azure, run `helpDocMd` for the API reference, and then `mkdocs` to build and publish the website.

Fixes #14 (though maybe not /well/). Also, the bulk of the work is thanks to @rtopfer's https://github.com/shimming-toolbox/helpDocMd which @po09i already worked on importing so I'm not at all taking credit for this.

kousu committed:
Port mkdocs into Azure.

* Add GH_PAGES_TOKEN (added as a secret on Azure, manually)

Also, isolate this pipeline to only run on changes to master.

Also, actually tries running the thing.

kousu commented:
I added a [github access token](https://github.com/settings/tokens) (with "repo" permission) as a [Azure secret](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#environment-variables) to this pipeline -- that's `$GH_PAGES_TOKEN`, and it won't work without it.

kousu commented:
Too bad this doesn't seem to run yet! For one thing, it crashes with a syntax error after trying to parse an [`arguments` block](https://www.mathworks.com/help/matlab/ref/arguments.html), a feature only added last year, because our self-hosted runner's Matlab is too old (which I think you figured out, I saw it mentioned in Slack).

It also still needs the actual `gh-pages` deployment ported from

* https://github.com/mhausenblas/mkdocs-deploy-gh-pages/blob/master/action.sh (`mkdocs gh-deploy`, which I assume is more or less doing this: https://github.com/DavidS/jekyll-deploy/blob/master/entrypoint.rb)

kousu commented:
Also I totally cancelled the current pipeline run. It's too bad that Github isn't picking that fact up. Maybe it will have figured it out by the morning.

kousu commented:
Also maybe this should be a "Deployment" not a "Pipeline"? I'm unclear on the difference.

po09i commented:
@kousu We are trying to set up the website on the other repo so that we can eventually merge this repo into shimming-toolbox. The script is currently being updated in this [PR](https://github.com/shimming-toolbox/shimming-toolbox/pull/126). If this works it would facilitate the other repo's implementation.
Just seeing this now

kousu commented:
Alright, I ported https://github.com/mhausenblas/mkdocs-deploy-gh-pages in. It's a bit messy because I didn't want to completely replace the ORIGIN, and I don't know yet what it's going to do for `git config user.name` on our self-hosted CI server, but it should build the website.

kousu commented:
Oh dear. I thought I told this to only run automatically on merges to master but it's running on the PR too, so the website is going to get overwritten *just for a PR*. That's a mistake.

kousu committed:
install mkdocs in a venv

(tristano has old versions of packages cached and this is throwing off pip, so disable the cache too).

kousu commented on generate-docs.sh:
```diff
@@ -0,0 +1,51 @@
+    #git config http."https://github.com/".extraheader "Authorization: token $GH_PAGES_TOKEN" # authenticate over HTTPS with the given token
+    # ^ this doesn't work. github will accept 'token ...' for API calls, but not for git+https://
+    # so instead we have to generate an HTTP basic auth:
+    auth="$(echo -n "x-access-token:$GH_PAGES_TOKEN" | base64 | tr -d '\r' | tr -d '\n')"
+    git config http."https://github.com/".extraheader "Authorization: basic "$auth"" # authenticate over HTTPS with the given token
+
+    git config url."https://github".insteadOf "git@github.com:"  # force HTTPS, not SSH
```
Auth is hard.

Computing the base64 in shell like this is unfortunate. I don't like it but it's the best I've got so far.

[Microsoft suggests](https://cloudblogs.microsoft.com/opensource/2019/04/05/publishing-github-pages-from-azure-pipelines/) using [Github Deploy Keys](https://developer.github.com/v3/guides/managing-deploy-keys/#deploy-keys). but that seems even harder.

The Action I ported from, https://github.com/mhausenblas/mkdocs-deploy-gh-pages, thought the thing to do was to replace the URL entirely with one with `x-access-token:$GH_PAGES_TOKEN` in it, and let the HTTP library handle the base64 step. But, not intending to run this on Github's infra, I don't have a reliable way to get the repo URLs besides the ones that come with the repo, so I would rather not try overwriting them.

Github Actions deploys like [this](https://github.com/DavidS/jekyll-deploy/blob/master/entrypoint.rb) benefit from the [built in logic](https://github.com/actions/checkout/blob/master/src/git-auth-helper.ts#L53-L66) which does exactly this algorithm before passing the repo off to the CI script.

I tried passing so many times `Authorization: token $GH_PAGES_TOKEN` which should have worked -- it worked for accessing api.github.com: https://github.com/DavidS/jekyll-deploy/issues/7 -- but I couldn't get it to go.

kousu replied:
It turns out http.extraHeader doesn't even work on our CI machine anyway: its git is too old. So I did it the jekyll-deploy way afterall.

kousu commented on azure-pipelines.yml:
```diff
@@ -1,15 +1,18 @@
 
 # TODO: push on master shimming-toolbox
-#trigger:
+trigger:
+- master
```
This is triggering on PRs and on commits, which is bad: I want the website to only be rebuilt when master is merged.

Docs here https://docs.microsoft.com/en-us/azure/devops/pipelines/repos/github?view=azure-devops&tabs=yaml#ci-triggers -- I don't know what I misread but I misread something.

kousu committed:
Embed the password in the URL afterall.

tristano has an older git that doesn't understand http.extraHeader.

kousu committed:
Change a word to see if the bot really updates the website

kousu commented:
@po09i minus the other issues in my comments this is working. https://github.com/shimming-toolbox/shimming-toolbox.org/commit/04584153546f0b68f3ceaf804ba5098d7aa52645 was made by the bot on `tristano` (notice it's by `agah_local`, not any of us).

You can copy it wholesale into https://github.com/shimming-toolbox/shimming-toolbox/pull/126 I think.

kousu commented on generate-docs.sh:
```diff
@@ -0,0 +1,57 @@
+# TODO: broken.
+#  1. writes to temp/shimming-toolbox/docs instead of docs/
+#  2. incompatible with Matlab R2019a which is what's on our CI machine.
+#matlab -nodisplay -nosplash -r "run('genDoc/generate_doc.m');exit"
```
**big caveat**: this doesn't actually run matlab yet. Matlab works on this environment, but the code we need to run on it is not compatible with the version ♈ 

Either matlab needs upgrading or the code needs some scalpels taken to it.

By the way, that script has a lot of assumptions. What I would like it to be is:

```
gen_matlab_api_doc(path_to_code_input, path_to_markdown_output)
```

jcohenadad replied:
as per our slack discussion, Matlab will be upgraded today by @alexfoias

po09i commented:
@kousu Thanks for all the work on this, website is currently generated automatically on every merge to master. This can be closed.
